app/src/main/java/com/aconno/blesniffer/adapter/DeserializerEditorAdapter.kt

```diff
@@ -1,142 +1,210 @@
         }
 
         private val converterTypeAdapter: ArrayAdapter<String> by lazy {
-            ArrayAdapter<String>(
-                    view.context,
-                    android.R.layout.simple_spinner_dropdown_item,
-                    ValueConverter.values().map { it.name }
+            ArrayAdapter(
+                view.context,
+                android.R.layout.simple_spinner_dropdown_item,
+                ValueConverter.values().map { it.name }
             ).apply {
                 setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
             }
         }
 
         init {
-            view.name.editText?.addTextChangedListener(this)
-            view.start.editText?.addTextChangedListener(this)
-            view.end.editText?.addTextChangedListener(this)
-            view.converter_type.onItemSelectedListener = this
-            view.remove.setOnClickListener(this)
-            view.color_button.setOnClickListener(this)
-            view.converter_type.adapter = converterTypeAdapter
+            registerTextWatchers()
+
+            view.spinner_converter_type.adapter = converterTypeAdapter
+            view.spinner_converter_type.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
+                override fun onItemSelected(parent: AdapterView<*>, v: View, position: Int, id: Long) {
+                    ValueConverter.values()[position].let { valueConverter ->
+                        fieldDeserializers[adapterPosition].type = valueConverter
+
+                        if (valueConverter.converter.length == -1) {
+                            view.til_end.isEnabled = true
+                        } else {
+                            view.til_end.isEnabled = false
+
+
+                            val value = Integer.parseInt(
+                                view.til_start.editText?.text?.toString()?.takeIf {
+                                    it.isNotEmpty()
+                                } ?: "0"
+                            )
+
+                            (value + valueConverter.converter.length).let { endIndexExclusive ->
+                                view.til_end.editText?.setText(endIndexExclusive.toString())
+                                fieldDeserializers[adapterPosition].endIndexExclusive = endIndexExclusive
+                            }
+                        }
+                    }
+                }
+
+                override fun onNothingSelected(p0: AdapterView<*>?) {}
+            }
+            view.btn_remove.setOnClickListener {
+                fieldDeserializers.removeAt(adapterPosition)
+                notifyItemRemoved(adapterPosition)
+            }
+            view.color_button.setOnClickListener {
+                showColorPickerDialog()
+            }
         }
 
+        private fun registerTextWatchers() {
+            view.til_name.editText?.addTextChangedListener(object : TextWatcher {
```

Comment:
You have a duplication of anonymous class declarations. You should define something like this:
```
AfterTextChangedWatcher(private val afterTextChanged: (Editable?) -> Unit ) : TextWatcher { ... }
```